tests/test_response.py

```diff
@@ -0,0 +1,149 @@
+    ],
+)
+def test_expiration(expire_after, is_expired, mock_session):
+    response = CachedResponse(mock_session.get(MOCKED_URL), expire_after)
+    assert response.from_cache is True
+    assert response.is_expired == is_expired
+
+
+def test_history(mock_session):
+    original_response = mock_session.get(MOCKED_URL)
+    original_response.history = [mock_session.get(MOCKED_URL)] * 3
+    response = CachedResponse(original_response)
+    assert len(response.history) == 3
+    assert all([isinstance(r, CachedResponse) for r in response.history])
+
+
+def test_raw_response__read(mock_session):
+    response = CachedResponse(mock_session.get(MOCKED_URL))
+    assert isinstance(response.raw, CachedHTTPResponse)
+    assert response.raw.read(10) == b'mock respo'
+    assert response.raw.read(None) == b'nse'
+    assert response.raw.read(1) == b''
+    assert response.raw._fp.closed
+
+
+def test_raw_response__reset(mock_session):
+    response = CachedResponse(mock_session.get(MOCKED_URL))
+    response.raw.read(None)
+    assert response.raw.read(1) == b''
+    assert response.raw._fp.closed
+
+    response.reset()
+    assert response.raw.read(None) == b'mock response'
+
+
+def test_raw_response__decode(mock_session):
+    """Test that a gzip-compressed raw response can be manually uncompressed with decode_content"""
+    url = f'{MOCKED_URL}/utf-8'
+    mock_session.mock_adapter.register_uri(
+        'GET',
+        url,
+        status_code=200,
+        body=BytesIO(gzip.compress(b'compressed response')),
+        headers={'content-encoding': 'gzip'},
+    )
+    response = CachedResponse(mock_session.get(url))
+    # Requests will have already read this, but let's just pretend we want to do it manually
+    response.raw._fp = BytesIO(gzip.compress(b'compressed response'))
+    assert response.raw.read(None, decode_content=True) == b'compressed response'
+
+
+def test_raw_response__stream(mock_session):
+    response = CachedResponse(mock_session.get(MOCKED_URL))
+    data = b''
+    for chunk in response.raw.stream(1):
+        data += chunk
+    assert data == b'mock response'
+    assert response.raw._fp.closed
+
+
+def test_raw_response__iterator(mock_session):
+    # Set up mock response with streamed content
+    url = f'{MOCKED_URL}/stream'
+    mock_raw_response = HTTPResponse(
+        body=BytesIO(b'mock response'),
+        status=200,
+        request_method='GET',
+        decode_content=False,
+        preload_content=False,
+    )
+    mock_session.mock_adapter.register_uri(
+        'GET',
+        url,
+        status_code=200,
+        raw=mock_raw_response,
+    )
+
+    # Expect the same chunks of data from the original response and subsequent cached responses
+    last_request_chunks = None
+    for i in range(3):
+        response = mock_session.get(url, stream=True)
+        chunks = list(response.iter_lines())
+        if i == 0:
+            assert response.from_cache is False
```

Comment:
as above, the asserts here could probably be simplified to `assert response.from_cache` and `assert not response.from_cache`. But I guess that's a matter of taste.

Reply:
Probably 80% of the time it doesn't make a difference, but explicitly checking boolean values can catch a few categories of bugs, or at least catch an error in the test.

By far the most common is checking things that are mock values when you don't expect them to be. For example, `unittest.Mock` will happily return a truthy value for any member. So if, say, `response` here was a mock object, and I made a typo below, `assert response.from_caaache` would pass, but `assert response.from_caaache is True` would not. On at least a couple occasions I've missed some bugs that made to to production because of that!